columnflow/tasks/ml.py

```diff
@@ -64,12 +70,29 @@ def __init__(self, *args, **kwargs):
                 f"{self.__class__.__name__}",
             )
 
+    @property
+    def preparation_producer_inst(self):
+        producer = self.ml_model_inst.preparation_producer(self.config_inst)
+        if producer and self._preparation_producer_inst is None:
```

Comment:
In case the `preparation_producer` method is doing something complicated, I would evaluate the method (first line) only when the `_preparation_producer_inst` is not yet cached.

Also, with the comment above, check for `law.no_value` instead of *None*.